File: frontend_web/src/services/gelisimimService.ts
import api from './api'

// ─── Tipler (mobil ile uyumlu) ───────────────────────────────────────────────

export interface GelisimimStats {
  completedTasks: number
  totalMinutes: number
  totalQuestions: number
  restDays: number
}

export interface XpInfo {
  totalXP: number
  currentLevelXP: number
  nextLevelXP: number
  levelName: string
  levelEmoji: string
  streakDays: number
  totalQuestions: number
}

export interface LessonDistribution {
  lessonName: string
  totalQuestions: number
}

export interface DailyQuestion {
  subjectName: string
  count: number
}

export interface DailyReport {
  date: string
  questions: DailyQuestion[]
  tasks: { completed: number; missed: number; totalMinutes: number }
  isEmpty: boolean
}

// ─── XP seviye hesabı (mobildeki applyXpBoost ile aynı) ──────────────────────

export function levelFromXp(total: number): {
  levelName: string
  levelEmoji: string
  currentLevelXP: number
  nextLevelXP: number
} {
  if (total <= 2000) return { levelName: 'Çırak Öğrenci', levelEmoji: '🌱', currentLevelXP: 0, nextLevelXP: 2000 }
  if (total <= 5000) return { levelName: 'Acemi Öğrenci', levelEmoji: '📖', currentLevelXP: 2000, nextLevelXP: 5000 }
  if (total <= 10000) return { levelName: 'Gelişen Öğrenci', levelEmoji: '📚', currentLevelXP: 5000, nextLevelXP: 10000 }
  return { levelName: 'Uzman Öğrenci', levelEmoji: '🎓', currentLevelXP: 10000, nextLevelXP: 20000 }
}

export function xpProgressFraction(xp: XpInfo): number {
  const range = xp.nextLevelXP - xp.currentLevelXP
  if (range <= 0) return 1
  return Math.min(1, Math.max(0, (xp.totalXP - xp.currentLevelXP) / range))
}

/** Backend XP'sine lokal tamamlanan ders XP'sini ekleyip seviyeyi yeniden hesaplar. */
export function applyXpBoost(base: XpInfo, boost: number): XpInfo {
  const total = base.totalXP + boost
  const lvl = levelFromXp(total)
  return {
    totalXP: total,
    currentLevelXP: lvl.currentLevelXP,
    nextLevelXP: lvl.nextLevelXP,
    levelName: lvl.levelName,
    levelEmoji: lvl.levelEmoji,
    streakDays: base.streakDays,
    totalQuestions: base.totalQuestions,
  }
}

// ─── Servis ───────────────────────────────────────────────────────────────────

export const gelisimimService = {
  async getStats(filter: 'all' | 'today'): Promise<GelisimimStats> {
    const res = await api.get(`/Gelisimim/stats?filter=${filter}`)
    const d = res.data ?? {}
    return {
      completedTasks: d.completedTasks ?? 0,
      totalMinutes: d.totalMinutes ?? 0,
      totalQuestions: d.totalQuestions ?? 0,
      restDays: d.restDays ?? 0,
    }
  },

  async getXpInfo(): Promise<XpInfo> {
    const res = await api.get('/Gelisimim/xp-info')
    const d = res.data ?? {}
    return {
      totalXP: d.totalXP ?? 0,
      currentLevelXP: d.currentLevelXP ?? 0,
      nextLevelXP: d.nextLevelXP ?? 2000,
      levelName: d.levelName ?? 'Çırak Öğrenci',
      levelEmoji: d.levelEmoji ?? '🌱',
      streakDays: d.streakDays ?? 0,
      totalQuestions: d.totalQuestions ?? 0,
    }
  },

  async getLessonDistribution(filter: 'all' | 'today'): Promise<LessonDistribution[]> {
    const res = await api.get(`/Gelisimim/lesson-distribution?filter=${filter}`)
    return ((res.data ?? []) as LessonDistribution[]).map((l) => ({
      lessonName: l.lessonName ?? '',
      totalQuestions: l.totalQuestions ?? 0,
    }))
  },

  async saveQuestions(entries: { subjectKey: string; subjectName: string; count: number }[]): Promise<number> {
    const res = await api.post('/Gelisimim/save-questions', { entries })
    return res.data?.totalToday ?? 0
  },

  async getCalendarActiveDays(year: number, month: number): Promise<string[]> {
    const res = await api.get(`/Gelisimim/calendar?year=${year}&month=${month}`)
    return (res.data?.activeDays ?? []) as string[]
  },

  /**
   * Bugün her ders için girilmiş soru sayısı: { subjectName: count }.
   * Anahtar yerine ders ADI kullanılır — mobil ve web farklı subjectKey
   * üretebildiği için (ör. 'tyt_coğrafya' vs 'tyt_cografya') ada göre
   * eşleştirmek tek güvenilir yoldur.
   */
  async getTodayQuestionCounts(): Promise<Record<string, number>> {
    const today = new Date()
    const dateStr = `${today.getFullYear()}-${String(today.getMonth() + 1).padStart(2, '0')}-${String(today.getDate()).padStart(2, '0')}`
    const report = await this.getDailyReport(dateStr)
    const map: Record<string, number> = {}
    for (const q of report.questions) {
      if ((q.count ?? 0) > 0) map[q.subjectName] = q.count
    }
    return map
  },

  async getDailyReport(date: string): Promise<DailyReport> {
    const res = await api.get(`/Gelisimim/daily-report?date=${date}`)
    const d = res.data ?? {}
    return {
      date: d.date ?? date,
      questions: (d.questions ?? []) as DailyQuestion[],
      tasks: {
        completed: d.tasks?.completed ?? 0,
        missed: d.tasks?.missed ?? 0,
        totalMinutes: d.tasks?.totalMinutes ?? 0,
      },
      isEmpty: d.isEmpty ?? true,
    }
  },
}

File: frontend_mobile/lib/services/gelisimim_service.dart
import 'api_service.dart';

class GelisimimStats {
  final int completedTasks;
  final int totalMinutes;
  final int totalQuestions;
  final int restDays;

  const GelisimimStats({
    required this.completedTasks,
    required this.totalMinutes,
    required this.totalQuestions,
    required this.restDays,
  });

  factory GelisimimStats.fromJson(Map<String, dynamic> json) => GelisimimStats(
        completedTasks: json['completedTasks'] ?? 0,
        totalMinutes: json['totalMinutes'] ?? 0,
        totalQuestions: json['totalQuestions'] ?? 0,
        restDays: json['restDays'] ?? 0,
      );

  static const empty = GelisimimStats(
    completedTasks: 0,
    totalMinutes: 0,
    totalQuestions: 0,
    restDays: 0,
  );
}

class SubjectEntry {
  final String key;
  final String name;
  final String icon;
  final int todayCount;
  int pendingCount; // local edit state (not persisted until Save)

  SubjectEntry({
    required this.key,
    required this.name,
    required this.icon,
    required this.todayCount,
    this.pendingCount = 0,
  });

  factory SubjectEntry.fromJson(Map<String, dynamic> json) => SubjectEntry(
        key: json['key'] ?? '',
        name: json['name'] ?? '',
        icon: json['icon'] ?? '📚',
        todayCount: json['todayCount'] ?? 0,
      );

  int get effectiveCount => pendingCount > 0 ? pendingCount : todayCount;
  bool get hasData => effectiveCount > 0;
}

class DailyQuestion {
  final String subjectName;
  final int count;

  const DailyQuestion({required this.subjectName, required this.count});

  factory DailyQuestion.fromJson(Map<String, dynamic> json) => DailyQuestion(
        subjectName: json['subjectName'] ?? '',
        count: json['count'] ?? 0,
      );
}

class DailyTasks {
  final int completed;
  final int missed;
  final int totalMinutes;

  const DailyTasks({
    required this.completed,
    required this.missed,
    required this.totalMinutes,
  });

  factory DailyTasks.fromJson(Map<String, dynamic> json) => DailyTasks(
        completed: json['completed'] ?? 0,
        missed: json['missed'] ?? 0,
        totalMinutes: json['totalMinutes'] ?? 0,
      );
}

class DailyReport {
  final String date;
  final List<DailyQuestion> questions;
  final DailyTasks tasks;
  final bool isEmpty;

  const DailyReport({
    required this.date,
    required this.questions,
    required this.tasks,
    required this.isEmpty,
  });

  factory DailyReport.fromJson(Map<String, dynamic> json) => DailyReport(
        date: json['date'] ?? '',
        questions: (json['questions'] as List? ?? [])
            .map((e) => DailyQuestion.fromJson(e as Map<String, dynamic>))
            .toList(),
        tasks: DailyTasks.fromJson(
            (json['tasks'] as Map<String, dynamic>?) ?? {}),
        isEmpty: json['isEmpty'] ?? true,
      );
}

class XpInfo {
  final int totalXP;
  final int currentLevelXP;
  final int nextLevelXP;
  final String levelName;
  final String levelEmoji;
  final int streakDays;
  final int streakBeforeToday;
  final int totalQuestions;

  const XpInfo({
    required this.totalXP,
    required this.currentLevelXP,
    required this.nextLevelXP,
    required this.levelName,
    required this.levelEmoji,
    required this.streakDays,
    required this.streakBeforeToday,
    required this.totalQuestions,
  });

  factory XpInfo.fromJson(Map<String, dynamic> json) => XpInfo(
        totalXP: json['totalXP'] ?? 0,
        currentLevelXP: json['currentLevelXP'] ?? 0,
        nextLevelXP: json['nextLevelXP'] ?? 2000,
        levelName: json['levelName'] ?? 'Çırak Öğrenci',
        levelEmoji: json['levelEmoji'] ?? '🌱',
        streakDays: json['streakDays'] ?? 0,
        streakBeforeToday: json['streakBeforeToday'] ?? 0,
        totalQuestions: json['totalQuestions'] ?? 0,
      );

  static const empty = XpInfo(
    totalXP: 0,
    currentLevelXP: 0,
    nextLevelXP: 2000,
    levelName: 'Çırak Öğrenci',
    levelEmoji: '🌱',
    streakDays: 0,
    streakBeforeToday: 0,
    totalQuestions: 0,
  );

  double get progressFraction {
    final range = nextLevelXP - currentLevelXP;
    if (range <= 0) return 1.0;
    return ((totalXP - currentLevelXP) / range).clamp(0.0, 1.0);
  }
}

class LessonDistribution {
  final String lessonName;
  final int totalQuestions;

  const LessonDistribution(
      {required this.lessonName, required this.totalQuestions});

  factory LessonDistribution.fromJson(Map<String, dynamic> json) =>
      LessonDistribution(
        lessonName: json['lessonName'] ?? '',
        totalQuestions: json['totalQuestions'] ?? 0,
      );
}

class GelisimimService {
  final _api = ApiService();

  Future<GelisimimStats> getStats(String filter) async {
    final res = await _api.dio
        .get('/Gelisimim/stats', queryParameters: {'filter': filter});
    return GelisimimStats.fromJson(res.data as Map<String, dynamic>);
  }

  Future<List<SubjectEntry>> getQuestionSubjects() async {
    final res = await _api.dio.get('/Gelisimim/question-subjects');
    return (res.data as List)
        .map((e) => SubjectEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<int> saveQuestions(List<SubjectEntry> entries) async {
    final body = {
      'entries': entries
          .where((e) => e.pendingCount > 0)
          .map((e) => {
                'subjectKey': e.key,
                'subjectName': e.name,
                'count': e.pendingCount,
              })
          .toList(),
    };
    final res = await _api.dio.post('/Gelisimim/save-questions', data: body);
    return (res.data as Map<String, dynamic>)['totalToday'] ?? 0;
  }

  Future<List<String>> getCalendarActiveDays(int year, int month) async {
    final res = await _api.dio.get(
      '/Gelisimim/calendar',
      queryParameters: {'year': year, 'month': month},
    );
    return ((res.data as Map<String, dynamic>)['activeDays'] as List)
        .cast<String>();
  }

  Future<DailyReport> getDailyReport(String date) async {
    final res = await _api.dio
        .get('/Gelisimim/daily-report', queryParameters: {'date': date});
    return DailyReport.fromJson(res.data as Map<String, dynamic>);
  }

  Future<XpInfo> getXpInfo() async {
    final res = await _api.dio.get('/Gelisimim/xp-info');
    return XpInfo.fromJson(res.data as Map<String, dynamic>);
  }

  Future<List<LessonDistribution>> getLessonDistribution(String filter) async {
    final res = await _api.dio.get('/Gelisimim/lesson-distribution',
        queryParameters: {'filter': filter});
    return (res.data as List)
        .map((e) => LessonDistribution.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}
